### PR TITLE
Issue #6774: Add UnnecessarySemicolonInTryWithResourcesCheck

### PR DESCRIPTION
**Pull request by strkkk**
#6774

Contribution repo PR  https://github.com/checkstyle/contribution/pull/378

Reg. report https://strkkk.github.io/6774/index.html

### CODE REVIEW

**Comment by strkkk** [comment]
@romani / @rnveach please take a look at this PR.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessarySemicolonInTryWithResourcesCheck.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+/**
+ * <p>
+ * Checks if unnecessary semicolon is used in last resource declaration.
+ * </p>
+ * <ul>
+ * <li>
+ * Property {@code allowWhenNoBraceAfterSemicolon} -
+ * Allow unnecessary semicolon if closing paren is not on the same line.
+ * Default value is {@code false}.
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessarySemicolonInTryWithResourcesCheck.java]: we need to update issue, looks like we lost this during final review of design.

default value should be "true".
as I want cases like:
```
          try(Reader r6 = new PipedReader();
              Reader r7 = new PipedReader(); // NO violation
          ){}
```
As it is friendly for minimization of diff.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessarysemicolonintrywithresources/InputUnnecessarySemicolonInTryWithResourcesAllow.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package com.puppycrawl.tools.checkstyle.checks.coding.unnecessarysemicolonintrywithresources;
+
+import java.io.IOException;
+import java.io.PipedReader;
+import java.io.Reader;
+
+public class InputUnnecessarySemicolonInTryWithResourcesAllow {
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessarysemicolonintrywithresources/InputUnnecessarySemicolonInTryWithResourcesAllow.java]: in new generation of inputs, we try to follow good practice - input file is for single configuration.
you did all correctly, but ...

Please add javadoc comment
```
/**
* "allowWhenNoBraceAfterSemicolon" ="false"
*
*/
```

one day in future UTs will take config from input files, and make sure that all "violation" trailing comments are matching to what we have in UT.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessarysemicolonintrywithresources/InputUnnecessarySemicolonInTryWithResources.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package com.puppycrawl.tools.checkstyle.checks.coding.unnecessarysemicolonintrywithresources;
+
+import java.io.IOException;
+import java.io.PipedReader;
+import java.io.Reader;
+
+public class InputUnnecessarySemicolonInTryWithResources {
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessarysemicolonintrywithresources/InputUnnecessarySemicolonInTryWithResources.java]: Please add javadoc comment
```
/**
* Default config.
*/
```

File: src/xdocs/config_coding.xml [hunk 4]
```diff
@@ -4987,6 +4987,117 @@ enum NoSemicolon {
...
+    <section name="UnnecessarySemicolonInTryWithResources">
+      <p>Since Checkstyle 8.22</p>
+      <subsection name="Description" id="UnnecessarySemicolonInTryWithResources_Description">
+        <p>
+            Checks if unnecessary semicolon is used in last resource declaration.
+        </p>
+      </subsection>
+
+      <subsection name="Properties" id="UnnecessarySemicolonInTryWithResources_Properties">
+            <table>
```
Comment (romani) [inline src/xdocs/config_coding.xml]: please indent as +2 in whole new section.

Comment (strkkk) [reply src/xdocs/config_coding.xml]: done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessarysemicolonintrywithresources/InputUnnecessarySemicolonInTryWithResources.java (hunk 3 above)
Comment (strkkk) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessarysemicolonintrywithresources/InputUnnecessarySemicolonInTryWithResources.java]: done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessarysemicolonintrywithresources/InputUnnecessarySemicolonInTryWithResourcesAllow.java (hunk 2 above)
Comment (strkkk) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessarysemicolonintrywithresources/InputUnnecessarySemicolonInTryWithResourcesAllow.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessarySemicolonInTryWithResourcesCheck.java (hunk 1 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessarySemicolonInTryWithResourcesCheck.java]: done.

Original issue is updated, docs and javadoc as well.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessarySemicolonInTryWithResourcesCheck.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+     * @param allowWhenNoBraceAfterSemicolon a value to set.
+     */
+    public void setAllowWhenNoBraceAfterSemicolon(boolean allowWhenNoBraceAfterSemicolon) {
+        this.allowWhenNoBraceAfterSemicolon = allowWhenNoBraceAfterSemicolon;
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        final DetailAST closingParen = ast.getLastChild();
+        final DetailAST tokenBeforeLast = closingParen.getPreviousSibling();
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessarySemicolonInTryWithResourcesCheck.java]: tokenBeforeCloseParen

File: src/xdocs/config_coding.xml [hunk 6]
```diff
@@ -4987,6 +4987,117 @@ enum NoSemicolon {
       </subsection>
     </section>
 
+    <section name="UnnecessarySemicolonInTryWithResources">
+      <p>Since Checkstyle 8.22</p>
+      <subsection name="Description" id="UnnecessarySemicolonInTryWithResources_Description">
+        <p>
+            Checks if unnecessary semicolon is used in last resource declaration.
```
Comment (romani) [inline src/xdocs/config_coding.xml]: indentation is "2" for all.

One day we will do even for java code.

File: src/xdocs/config_coding.xml [hunk 7]
```diff
@@ -4987,6 +4987,117 @@ enum NoSemicolon {
...
+
+      <subsection name="Examples" id="UnnecessarySemicolonInTryWithResources_Examples">
+        <p>
+          To configure the check:
+        </p>
+        <source>
+&lt;module name=&quot;UnnecessarySemicolonInTryWithResources&quot;/&gt;
+        </source>
+        <p>
+           Example of violations
```
Comment (romani) [inline src/xdocs/config_coding.xml]: fix indentaion

File: src/xdocs/config_coding.xml [hunk 8]
```diff
@@ -4987,6 +4987,117 @@ enum NoSemicolon {
...
+          To configure the check to detect unnecessary semicolon
+          if closing paren is not on same line
+        </p>
+        <source>
+&lt;module name="UnnecessarySemicolonInTryWithResources"&gt;
+  &lt;property name="allowWhenNoBraceAfterSemicolon" value="false"/&gt;
+&lt;/module&gt;
+        </source>
+        <p>
+           Example of exclusion
```
Comment (romani) [inline src/xdocs/config_coding.xml]: fix

File: src/xdocs/config_coding.xml (hunk 7 above)
Comment (strkkk) [reply src/xdocs/config_coding.xml]: done

File: src/xdocs/config_coding.xml (hunk 8 above)
Comment (strkkk) [reply src/xdocs/config_coding.xml]: done

File: src/xdocs/config_coding.xml (hunk 6 above)
Comment (strkkk) [reply src/xdocs/config_coding.xml]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessarySemicolonInTryWithResourcesCheck.java (hunk 5 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessarySemicolonInTryWithResourcesCheck.java]: done

**Pull request merged**